### ISSUE TITLE
`tf.strided_slice()` option for `inclusive_end`

### ISSUE THREAD

**Issue by t-kalinowski** [tensorflow/tensorflow#55237]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): Latest (2.8)
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

I am maintainer of the [R interface to tensorflow](https://github.com/rstudio/tensorflow). The convention in R is for subsetting with `[` to be 1-based, and inclusive of the slice end. In most situations it is trivial to adapt the `[` tensor method in R to Python convention of 0-based indices and a half-open interval. Except for one case: when the slice end is also the first element of the container.

For example,
```
x = [1, 2, 3, 4, 5, 6]
```
An R call like `x[3:1]` would return `[3, 2, 1]`. To translate this to Python, the call would be `x[2::-1]. This translation is easy enough to do in eager mode, but it's difficult in graph mode if the user supplied the slice end as a tensor, then the type of `end` becomes varient. (because it can be an integer, or `None`, depending on the tensor value at run time.

Wrapping the `end` in a `tf.experimental.Optional()` doesn't work, nor does `tf.strided_slice()` accept a tensor value for the `end_mask()`. This means that there is no way to faithfully translate an R call like `x[start:end]` if `end` is a tensor and `end` could potentially resolve to the first element of the tensor.

**Will this change the current api? How?**

There are a couple of ways to resolve this, but the simplest I believe is to add an argument with default to `tf.strided_slice(..., inclusive_end=False)`. This would be no breaking change to existing code. From the R interface, we can then call `tf.strided_slice(..., inclusive_end = True)`.

**Who will benefit with this feature?**

All Tensorflow users from R, as well as interfaces to Tensorflow from other languages with 1-based and closed interval slice semantics (e.g., Julia).

**Any Other info.**

Having an additional arg `tf.strided_slice(..., one_based = False)` would be icing on the cake :)

**Comment by wangpengmit** [comment]
Thanks for identifying the problem! This is a reasonable feature request. Unfortunately I'm not sure we'll have bandwidth for it in the near term.

**Comment by t-kalinowski** [comment]
Thanks for responding. This would be a great quality-of-life improvement for R users, and I'd be willing to help how I can implement it. Would you be open to accepting a PR for this?

**Comment by wangpengmit** [comment]
Yeah I'm happy to accept a PR for this. Your contribution will be appreciated!

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.

**Comment by t-kalinowski** [comment]
Hi @tilakrayal, I had looked into implementing this earlier but didn’t have enough context (like knowledge of Bazel) to complete it in a reasonable time.

However, the feature for 1-based slicing would still be very helpful for interfaces from 1-based languages like R, Julia, etc. It would also be useful for creating new interfaces for tools like JAX or XLA.

Would your team be able to implement this without too much effort?
